Enforced directory structure for IDL files

Hello there, posting this more as a discussion rather than an issue.

I am facing an issue in my CMake project when compiling ROS 2 from sources, after having included this repo in my build workspace (it has been added to ros2.repos recently). In particular, I am invoking `rosidl_generate_interfaces`, passing an idl file within an `idl` folder. It looks like rosidl rust generator expects the file to be instead under a `msg` folder (and also use "module msg" in the definition). This restriction does not seem to be imposed by the other registered generators, so I am wondering if the condition should be relaxed/removed, or if you are actually following a standard or rule and hence I should be the one adapting to it.

Thanks in advance!

Hey thanks for the report. I just wanted to confirm this behavior is indeed what we currently have.

```
CMake Error at /home/sam/rust_ws/install/rosidl_generator_rs/share/rosidl_generator_rs/cmake/rosidl_generator_rs_generate_interfaces.cmake:50 (message):
  Interface file with unknown parent folder:
  /home/sam/rust_ws/src/ros2-rust/examples/rclrs/rclrs_example_msgs/idl/MyMessage.idl
Call Stack (most recent call first):
  /opt/ros/rolling/share/ament_cmake_core/cmake/core/ament_execute_extensions.cmake:48 (include)
  /opt/ros/rolling/share/rosidl_cmake/cmake/rosidl_generate_interfaces.cmake:280 (ament_execute_extensions)
  CMakeLists.txt:23 (rosidl_generate_interfaces)
```

Its possible this can be relaxed, but I/we will need to spend some time understanding how other generators handle these arbitrary paths. I suspect this work would be non-trivial for our generator. 

We've been out on the build farm for a while now, is this really that rare of a situation? I'm surprised no one has brought this up before.

---

> and also use "module msg" in the definition

Don't all of the generators require you to mimic the folder layout in modules?
```
/rclrs_example_msgs# tree -L 2
.
├── CMakeLists.txt
├── idl
│   └── MyMessage.idl
```
I would expect to see this in `MyMessage.idl`
```
module rclrs_example_msgs {
  module idl {
    ...
```

--- a/rosidl_generator_rs/cmake/custom_command.cmake
+++ b/rosidl_generator_rs/cmake/custom_command.cmake
@@ -18,9 +18,6 @@ set(PYTHON_EXECUTABLE ${Python3_EXECUTABLE})
 add_custom_command(
   OUTPUT
   ${_generated_common_rs_files}
-  ${_generated_msg_rs_files}
-  ${_generated_srv_rs_files}
-  ${_generated_action_rs_files}
   COMMAND ${PYTHON_EXECUTABLE} ${rosidl_generator_rs_BIN}
   --generator-arguments-file "${generator_arguments_file}"
   --typesupport-impls "${_typesupport_impls}"
@@ -36,8 +33,5 @@ else()
     ${rosidl_generate_interfaces_TARGET}${_target_suffix} ALL
     DEPENDS
     ${_generated_common_rs_files}
-    ${_generated_msg_rs_files}
-    ${_generated_srv_rs_files}
-    ${_generated_action_rs_files}
   )
 endif()

--- a/rosidl_generator_rs/cmake/rosidl_generator_rs_generate_interfaces.cmake
+++ b/rosidl_generator_rs/cmake/rosidl_generator_rs_generate_interfaces.cmake
@@ -16,31 +16,8 @@ set(_output_path
   "${CMAKE_CURRENT_BINARY_DIR}/rosidl_generator_rs/${PROJECT_NAME}")
 set(_generated_common_rs_files "")
 
-set(_generated_msg_rs_files "")
-set(_generated_srv_rs_files "")
-set(_generated_action_rs_files "")
-
-set(_has_msg FALSE)
-set(_has_srv FALSE)
-set(_has_action FALSE)
-
 foreach(_idl_file ${rosidl_generate_interfaces_ABS_IDL_FILES})
-  get_filename_component(_parent_folder "${_idl_file}" DIRECTORY)
-  get_filename_component(_parent_folder "${_parent_folder}" NAME)
-  get_filename_component(_module_name "${_idl_file}" NAME_WE)
-
-  if(_parent_folder STREQUAL "msg")
-    set(_has_msg TRUE)
-    set(_idl_files ${_idl_files} ${_idl_file})
-  elseif(_parent_folder STREQUAL "srv")
-    set(_has_srv TRUE)
-    set(_idl_files ${_idl_files} ${_idl_file})
-  elseif(_parent_folder STREQUAL "action")
-    set(_has_action TRUE)
-    set(_idl_files ${_idl_files} ${_idl_file})
-  else()
-    message(FATAL_ERROR "Interface file with unknown parent folder: ${_idl_file}")
-  endif()
+  list(APPEND _idl_files ${_idl_file})
 endforeach()
 
 list(APPEND _generated_common_rs_files
@@ -48,27 +25,6 @@ list(APPEND _generated_common_rs_files
   "${_output_path}/rust/build.rs"
   "${_output_path}/rust/Cargo.toml"
 )
-
-if(${_has_msg})
-  list(APPEND _generated_msg_rs_files
-    "${_output_path}/rust/src/msg.rs"
-    "${_output_path}/rust/src/msg/rmw.rs"
-  )
-endif()
-
-if(${_has_srv})
-  list(APPEND _generated_srv_rs_files
-    "${_output_path}/rust/src/srv.rs"
-    "${_output_path}/rust/src/srv/rmw.rs"
-  )
-endif()
-
-if(${_has_action})
-  list(APPEND _generated_action_rs_files
-    "${_output_path}/rust/src/action.rs"
-    "${_output_path}/rust/src/action/rmw.rs"
-  )
-endif()
 
 set(_dependency_files "")
 set(_dependencies "")
@@ -86,6 +42,9 @@ set(target_dependencies
   ${rosidl_generator_rs_GENERATOR_FILES}
   "${rosidl_generator_rs_TEMPLATE_DIR}/action.rs.em"
   "${rosidl_generator_rs_TEMPLATE_DIR}/action/rmw.rs.em"
+  "${rosidl_generator_rs_TEMPLATE_DIR}/build.rs.em"
+  "${rosidl_generator_rs_TEMPLATE_DIR}/Cargo.toml.em"
+  "${rosidl_generator_rs_TEMPLATE_DIR}/lib.rs.em"
   "${rosidl_generator_rs_TEMPLATE_DIR}/msg.rs.em"
   "${rosidl_generator_rs_TEMPLATE_DIR}/msg/rmw.rs.em"
   "${rosidl_generator_rs_TEMPLATE_DIR}/srv.rs.em"
@@ -94,7 +53,6 @@ set(target_dependencies
   "${rosidl_generator_rs_TEMPLATE_DIR}/templates/msg_rmw.rs.em"
   "${rosidl_generator_rs_TEMPLATE_DIR}/templates/srv_idiomatic.rs.em"
   "${rosidl_generator_rs_TEMPLATE_DIR}/templates/srv_rmw.rs.em"
-  ${rosidl_generate_interfaces_ABS_IDL_FILES}
   ${_idl_files}
   ${_dependency_files})
 foreach(dep ${target_dependencies})
@@ -137,9 +95,6 @@ add_dependencies(${rosidl_generate_interfaces_TARGET} ${rosidl_generate_interfac
 set_property(
   SOURCE
   ${_generated_common_rs_files}
-  ${_generated_msg_rs_files}
-  ${_generated_srv_rs_files}
-  ${_generated_action_rs_files}
   PROPERTY GENERATED 1)
 
 set(_rsext_suffix "__rsext")
@@ -152,11 +107,7 @@ if(NOT rosidl_generate_interfaces_SKIP_INSTALL)
 endif()
 
 if(BUILD_TESTING AND rosidl_generate_interfaces_ADD_LINTER_TESTS)
-  if(
-    NOT _generated_msg_rs_files STREQUAL "" OR
-    NOT _generated_srv_rs_files STREQUAL "" OR
-    NOT _generated_action_rs_files STREQUAL ""
-  )
+  if(NOT _idl_files STREQUAL "")
   # TODO(esteve): add linters for Rust files
   endif()
 endif()

--- a/rosidl_generator_rs/resource/lib.rs.em
+++ b/rosidl_generator_rs/resource/lib.rs.em
@@ -2,29 +2,12 @@
 #![allow(clippy::derive_partial_eq_without_eq)]
 #![allow(clippy::upper_case_acronyms)]
 
-@[if len(msg_specs) > 0]@
-#[path = "msg.rs"]
-mod msg_idiomatic;
-pub mod msg {
-    pub use super::msg_idiomatic::*;
+@[for namespace in generated_namespaces]@
+#[path = "@(namespace).rs"]
+mod @(namespace)_idiomatic;
+pub mod @(namespace) {
+    pub use super::@(namespace)_idiomatic::*;
     pub mod rmw;
 }
-@[end if]@
 
-@[if len(srv_specs) > 0]@
-#[path = "srv.rs"]
-mod srv_idiomatic;
-pub mod srv {
-    pub use super::srv_idiomatic::*;
-    pub mod rmw;
-}
-@[end if]@
-
-@[if len(action_specs) > 0]@
-#[path = "action.rs"]
-mod action_idiomatic;
-pub mod action {
-    pub use super::action_idiomatic::*;
-    pub mod rmw;
-}
-@[end if]@
+@[end for]@

--- a/rosidl_generator_rs/rosidl_generator_rs/__init__.py
+++ b/rosidl_generator_rs/rosidl_generator_rs/__init__.py
@@ -64,6 +64,88 @@ def convert_lower_case_underscore_to_camel_case(word):
     return ''.join(x.capitalize() or '_' for x in word.split('_'))
 
 
+def _namespace_from_namespaced_type(namespaced_type):
+    namespaces = list(namespaced_type.namespaces)
+    if not namespaces or namespaces[0] != package_name:
+        raise ValueError(
+            f"Expected namespace to start with package '{package_name}', got {namespaces}")
+    if len(namespaces) != 2:
+        raise ValueError(
+            f"Expected exactly one namespace component after package '{package_name}', got {namespaces}")
+    namespace = namespaces[1]
+    if not namespace.isidentifier() or get_rs_name(namespace) != namespace:
+        raise ValueError(
+            f"Namespace '{namespace}' cannot be emitted as a Rust module name")
+    return namespace
+
+
+def _namespace_from_message(message):
+    return _namespace_from_namespaced_type(message.structure.namespaced_type)
+
+
+def _namespace_from_service(service):
+    return _namespace_from_namespaced_type(service.namespaced_type)
+
+
+def _namespace_from_action(action):
+    return _namespace_from_namespaced_type(action.namespaced_type)
+
+
+def _group_specs_by_namespace(specs, namespace_getter):
+    grouped = {}
+    for spec in specs:
+        namespace = namespace_getter(spec)
+        grouped.setdefault(namespace, []).append(spec)
+    return grouped
+
+
+def _validate_single_kind_per_namespace(namespace_to_kinds):
+    for namespace, kinds in namespace_to_kinds.items():
+        direct_kinds = {kind for kind, _ in kinds}
+        if len(direct_kinds) > 1:
+            raise ValueError(
+                f"Namespace '{namespace}' contains multiple top-level kinds: {sorted(direct_kinds)}")
+
+
+def _expand_namespace_templates(template_dir, output_dir, namespace, spec_kind, specs,
+                                latest_target_timestamp, data):
+    if not specs:
+        return
+
+    if spec_kind == 'msg':
+        mappings = {
+            os.path.join(template_dir, 'msg.rs.em'): [f'rust/src/{namespace}.rs'],
+            os.path.join(template_dir, 'msg/rmw.rs.em'): [f'rust/src/{namespace}/rmw.rs'],
+        }
+        template_specs = 'msg_specs'
+    elif spec_kind == 'srv':
+        mappings = {
+            os.path.join(template_dir, 'srv.rs.em'): [f'rust/src/{namespace}.rs'],
+            os.path.join(template_dir, 'srv/rmw.rs.em'): [f'rust/src/{namespace}/rmw.rs'],
+        }
+        template_specs = 'srv_specs'
+    elif spec_kind == 'action':
+        mappings = {
+            os.path.join(template_dir, 'action.rs.em'): [f'rust/src/{namespace}.rs'],
+            os.path.join(template_dir, 'action/rmw.rs.em'): [f'rust/src/{namespace}/rmw.rs'],
+        }
+        template_specs = 'action_specs'
+    else:
+        raise ValueError(f'Unknown spec kind {spec_kind}')
+
+    namespace_data = data.copy()
+    namespace_data[template_specs] = [(namespace, spec) for spec in specs]
+
+    for template_file, generated_filenames in mappings.items():
+        for generated_filename in generated_filenames:
+            generated_file = os.path.join(output_dir, generated_filename)
+            rosidl_pycommon.expand_template(
+                os.path.join(template_dir, template_file),
+                namespace_data.copy(),
+                generated_file,
+                minimum_timestamp=latest_target_timestamp)
+
+
 def generate_rs(generator_arguments_file, typesupport_impls):
     args = rosidl_pycommon.read_generator_arguments(generator_arguments_file)
 
@@ -99,31 +181,17 @@ def generate_rs(generator_arguments_file, typesupport_impls):
 
     template_dir = args['template_dir']
 
-    mapping_msgs = {
-        os.path.join(template_dir, 'msg.rs.em'): ['rust/src/%s'],
-        os.path.join(template_dir, 'msg/rmw.rs.em'): ['rust/src/msg/%s'],
-    }
-
-    mapping_srvs = {
-        os.path.join(template_dir, 'srv.rs.em'): ['rust/src/%s'],
-        os.path.join(template_dir, 'srv/rmw.rs.em'): ['rust/src/srv/%s'],
-    }
-
-    mapping_actions = {
-        os.path.join(template_dir, 'action.rs.em'): ['rust/src/%s'],
-        os.path.join(template_dir, 'action/rmw.rs.em'): ['rust/src/action/%s'],
-    }
-
     # Ensure the required templates exist
-    for template_file in mapping_msgs.keys():
+    for template_file in [
+        os.path.join(template_dir, 'msg.rs.em'),
+        os.path.join(template_dir, 'msg/rmw.rs.em'),
+        os.path.join(template_dir, 'srv.rs.em'),
+        os.path.join(template_dir, 'srv/rmw.rs.em'),
+        os.path.join(template_dir, 'action.rs.em'),
+        os.path.join(template_dir, 'action/rmw.rs.em'),
+    ]:
         assert os.path.exists(template_file), \
-            'Messages template file %s not found' % template_file
-    for template_file in mapping_srvs.keys():
-        assert os.path.exists(template_file), \
-            'Services template file %s not found' % template_file
-    for template_file in mapping_actions.keys():
-        assert os.path.exists(template_file), \
-            'Actions template file %s not found' % template_file
+            'Template file %s not found' % template_file
 
     data = {
         'pre_field_serde': pre_field_serde,
@@ -146,53 +214,44 @@ def generate_rs(generator_arguments_file, typesupport_impls):
     latest_target_timestamp = rosidl_pycommon.get_newest_modification_time(
         args['target_dependencies'])
 
-    for message in idl_content.get_elements_of_type(Message):
-        data['msg_specs'].append(('msg', message))
+    message_specs = list(idl_content.get_elements_of_type(Message))
+    service_specs = list(idl_content.get_elements_of_type(Service))
+    action_specs = list(idl_content.get_elements_of_type(Action))
 
-    for service in idl_content.get_elements_of_type(Service):
-        data['srv_specs'].append(('srv', service))
+    messages_by_namespace = _group_specs_by_namespace(
+        message_specs, _namespace_from_message)
+    services_by_namespace = _group_specs_by_namespace(
+        service_specs, _namespace_from_service)
+    actions_by_namespace = _group_specs_by_namespace(
+        action_specs, _namespace_from_action)
 
-    for action in idl_content.get_elements_of_type(Action):
-        data['action_specs'].append(('action', action))
+    namespace_to_kinds = {}
+    for namespace, specs in messages_by_namespace.items():
+        namespace_to_kinds.setdefault(namespace, []).append(('msg', specs))
+    for namespace, specs in services_by_namespace.items():
+        namespace_to_kinds.setdefault(namespace, []).append(('srv', specs))
+    for namespace, specs in actions_by_namespace.items():
+        namespace_to_kinds.setdefault(namespace, []).append(('action', specs))
 
-    if data['msg_specs']:
-        for template_file, generated_filenames in mapping_msgs.items():
-            stem = _removesuffix(Path(template_file).stem, ".em")
+    _validate_single_kind_per_namespace(namespace_to_kinds)
 
-            for generated_filename in generated_filenames:
-                generated_file = os.path.join(args['output_dir'],
-                                              generated_filename % stem)
-                rosidl_pycommon.expand_template(
-                    os.path.join(template_dir, template_file),
-                    data.copy(),
-                    generated_file,
-                    minimum_timestamp=latest_target_timestamp)
+    for namespace in sorted(namespace_to_kinds):
+        kinds = namespace_to_kinds[namespace]
+        kind, specs = kinds[0]
+        _expand_namespace_templates(
+            template_dir, args['output_dir'], namespace, kind, specs,
+            latest_target_timestamp, data)
 
-    if data['srv_specs']:
-        for template_file, generated_filenames in mapping_srvs.items():
-            stem = _removesuffix(Path(template_file).stem, ".em")
-
-            for generated_filename in generated_filenames:
-                generated_file = os.path.join(args['output_dir'],
-                                              generated_filename % stem)
-                rosidl_pycommon.expand_template(
-                    os.path.join(template_dir, template_file),
-                    data.copy(),
-                    generated_file,
-                    minimum_timestamp=latest_target_timestamp)
-
-    if data['action_specs']:
-        for template_file, generated_filenames in mapping_actions.items():
-            stem = _removesuffix(Path(template_file).stem, ".em")
-
-            for generated_filename in generated_filenames:
-                generated_file = os.path.join(args['output_dir'],
-                                              generated_filename % stem)
-                rosidl_pycommon.expand_template(
-                    os.path.join(template_dir, template_file),
-                    data.copy(),
-                    generated_file,
-                    minimum_timestamp=latest_target_timestamp)
+    data['generated_namespaces'] = sorted(namespace_to_kinds)
+    data['msg_specs'] = [
+        (_namespace_from_message(message), message)
+        for message in message_specs]
+    data['srv_specs'] = [
+        (_namespace_from_service(service), service)
+        for service in service_specs]
+    data['action_specs'] = [
+        (_namespace_from_action(action), action)
+        for action in action_specs]
 
     rosidl_pycommon.expand_template(
         os.path.join(template_dir, 'lib.rs.em'),

